barman/retention_policies.py

```diff
@@ -126,6 +126,59 @@ def to_json(self):
         """
         return "%s %s %s" % (self.mode, self.value, self.unit)
 
+    def propagate_annotation_to_children(self, backup_info, report, annotation):
+        """
+        Propagate annotation to all backups in the chain.
+        This has a side-effect. It modifies or add data to report dict.
+
+        :param barman.infofile.BackupInfo backup_info: The object we want to
+            propagate annotations from.
+        :param report dict[str, str]: The report data structure to be modified.
+        :param annotation str:  The status of the backup according to retention
+            policies
+        """
+        backup_id_tree = [
+            backup.backup_id for backup in list(backup_info.walk_backups_tree())
+        ][:-1]
+        for _id in backup_id_tree:
+            report[_id] = annotation
+        len_format = 2 if len(backup_id_tree) <= 7 else 3
+        msg = "Propagating %s annotation of backup %s to n=%s children %s." % (
+            annotation,
+            backup_info.backup_id,
+            str(len(backup_id_tree)),
+            ",".join(backup_id_tree[:len_format])
+            + "..."
+            + ",".join(backup_id_tree[-len_format:]),
+        )
+        _logger.debug(msg)
+
+    def has_children(self, backup_info):
+        """
+        Only checks if the backup_info object has children
+
+        :param barman.infofile.BackupInfo backup_info: The object we want to
+            do the check.
+        """
+        return (
+            isinstance(backup_info, BackupInfo)
+            and hasattr(backup_info, "children_backup_ids")
+            and backup_info.children_backup_ids is not None
+        )
+
+    def has_parent(self, backup_info):
```

Comment:
A method like this seems to fit best in `LocalBackupInfo`.
Maybe we could eliminate both `has_children` and `has_parent` and introduce a `LocalBackupInfo.is_incremental` method.